lib/linter/apply-disable-directives.js

```diff
@@ -197,62 +197,93 @@ function processUnusedDisableDirectives(allDirectives) {
  * for the exported function, except that `reportUnusedDisableDirectives` is not supported
  * (this function always reports unused disable directives).
  * @returns {{problems: Problem[], unusedDisableDirectives: Problem[]}} An object with a list
- * of filtered problems and unused eslint-disable directives
+ * of problems (including suppressed ones) and unused eslint-disable directives
  */
 function applyDirectives(options) {
     const problems = [];
-    let nextDirectiveIndex = 0;
     let currentGlobalDisableDirective = null;
-    const disabledRuleMap = new Map();
 
-    // enabledRules is only used when there is a current global disable directive.
-    const enabledRules = new Set();
     const usedDisableDirectives = new Set();
 
     for (const problem of options.problems) {
```

Comment:
This loop has a lot of code that made sense before this change, when it was optimized to go through directives only once. If we can't keep that optimization, then the loop can be much simpler:

```js
for (const problem of options.problems) {
    let disableDirectivesForProblem = [];
    let nextDirectiveIndex = 0;

    while (
        nextDirectiveIndex < options.directives.length &&
        compareLocations(options.directives[nextDirectiveIndex], problem) <= 0
    ) {
        const directive = options.directives[nextDirectiveIndex++];

        if (directive.ruleId === null || directive.ruleId === problem.ruleId) {
            switch (directive.type) {
                case "disable":
                    disableDirectivesForProblem.push(directive);
                    break;

                case "enable":
                    disableDirectivesForProblem = [];
                    break;

                // no default
            }
        }
    }

    if (disableDirectivesForProblem.length > 0) {
        const suppressions = disableDirectivesForProblem.map(directive => ({
            kind: "directive",
            justification: directive.unprocessedDirective.justification
        }));

        if (problem.suppressions) {
            problem.suppressions = problem.suppressions.concat(suppressions);
        } else {
            problem.suppressions = suppressions;
            usedDisableDirectives.add(disableDirectivesForProblem[disableDirectivesForProblem.length - 1]);
        }
    }

    problems.push(problem);
}
```

Reply:
Many thanks! I didn't realize the "used" directive should be the last working one.